### PR TITLE
Sort a few python dicts for deterministic builds, even with python 3.5 and earlier.

### PR DESCRIPTION
**Pull request by marc-hb**
See commit messages.

### CODE REVIEW

File: scripts/dts/extract_dts_includes.py [hunk 1]
```diff
@@ -441,7 +441,9 @@ def yaml_inc_error(msg):
 def generate_defines():
     # Generates #defines (and .conf file values) from DTS
 
-    for node_path in reduced.keys():
+    # sorted() otherwise Python < 3.6 randomizes the order of the flash
+    # partition table
+    for node_path in sorted(reduced.keys()):
```
Comment (ulfalizer) [inline scripts/dts/extract_dts_includes.py]: Looks like I missed this earlier, but could get rid of the `.keys()`. `x in dict`, `for x in dict`, `sorted(dict)`, etc., always uses the keys.

File: scripts/gen_kobject_list.py [hunk 2]
```diff
@@ -269,10 +276,14 @@ def main():
     parse_args()
 
     if args.gperf_output:
+        assert args.kernel, "--kernel ELF required for --gperf-output"
         eh = ElfHelper(args.kernel, args.verbose, kobjects, subsystems)
         syms = eh.get_symbols()
         max_threads = syms["CONFIG_MAX_THREAD_BYTES"] * 8
         objs = eh.find_kobjects(syms)
+        if len(objs) == 0:
```
Comment (ulfalizer) [inline scripts/gen_kobject_list.py]: Can do `if not objs:`.

File: scripts/gen_priv_stacks.py [hunk 3]
```diff
@@ -101,10 +101,14 @@ def parse_args():
 def main():
     parse_args()
 
+    assert args.kernel, "--kernel ELF required for --gperf-output"
     eh = ElfHelper(args.kernel, args.verbose, kobjects, [])
     syms = eh.get_symbols()
     max_threads = syms["CONFIG_MAX_THREAD_BYTES"] * 8
     objs = eh.find_kobjects(syms)
+    if len(objs) == 0:
```
Comment (ulfalizer) [inline scripts/gen_priv_stacks.py]: Ditto here.

File: scripts/dts/extract_dts_includes.py (hunk 1 above)
Comment (marc-hb) [reply scripts/dts/extract_dts_includes.py]: Agreed but I'd like to keep this minimal fix as minimal as possible :-) Not just for aesthetic reasons but also cherry-picking for testing / backporting etc.

File: scripts/gen_priv_stacks.py (hunk 3 above)
Comment (ulfalizer) [reply scripts/gen_priv_stacks.py]: Looks like `objs` isn't used after this point, so could make it `if not eh.find_kobjects(syms):` too. Makes it clear that the result isn't needed beyond that check.

File: scripts/gen_priv_stacks.py [hunk 4]
```diff
@@ -101,10 +101,14 @@ def parse_args():
 def main():
     parse_args()
 
+    assert args.kernel, "--kernel ELF required for --gperf-output"
```
Comment (ulfalizer) [inline scripts/gen_priv_stacks.py]: There's a `required=True` on the `parser.add_argument()` that adds `--kernel`, so is this needed?

File: scripts/gen_kobject_list.py [hunk 5]
```diff
@@ -269,10 +276,14 @@ def main():
     parse_args()
 
     if args.gperf_output:
+        assert args.kernel, "--kernel ELF required for --gperf-output"
```
Comment (ulfalizer) [inline scripts/gen_kobject_list.py]: Maybe

```python
if not args.kernel:
    sys.exit("--kernel must be passed if --gperf-output is")
```

would be a bit more purist.

It's more of a usage error than a runtime error.

**Review by ulfalizer — CHANGES_REQUESTED** [review]

File: scripts/dts/extract_dts_includes.py (hunk 1 above)
Comment (ulfalizer) [reply scripts/dts/extract_dts_includes.py]: Alright, fine with me.

File: scripts/gen_priv_stacks.py (hunk 4 above)
Comment (marc-hb) [reply scripts/gen_priv_stacks.py]: You're right, I saw the mostly duplicated `main()` so I blindly copied the assert from the other script. Will remove.

File: scripts/gen_priv_stacks.py (hunk 3 above)
Comment (marc-hb) [reply scripts/gen_priv_stacks.py]: Will do.

File: scripts/gen_kobject_list.py (hunk 5 above)
Comment (marc-hb) [reply scripts/gen_kobject_list.py]: It is a usage error but... a pretty stupid, "trial-and-error" one so I didn't want to spend too many lines of code on it.

File: scripts/gen_kobject_list.py (hunk 2 above)
Comment (marc-hb) [reply scripts/gen_kobject_list.py]: Will do.

**Review by tyrantlizard77 — APPROVED** [review]
other than the minor issues @ulfalizer noted this LGTM

**Review by nashif — APPROVED** [review]

**Review by ulfalizer — APPROVED** [review]
Still think the `assert` for reporting a usage error looks confusing, but won't block over it.

**Comment by marc-hb** [comment]
Thanks for the very prompt reviews. For the record this is a "screenshot" of a typical diff: https://gist.github.com/nashif/675dd70be922f0200c181d6a65baeafb

There's also one last `crypto.mbedtls/zephyr/include/generated/app_smem.ld` randomness there which no one has reproduced, maybe addressed elsewhere?

**Pull request merged**